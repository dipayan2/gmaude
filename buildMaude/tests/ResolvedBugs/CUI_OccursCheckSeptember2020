#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/CUI_OccursCheckSeptember2020.maude -no-banner -no-advise \
  > CUI_OccursCheckSeptember2020.out 2>&1

diff $srcdir/CUI_OccursCheckSeptember2020.expected CUI_OccursCheckSeptember2020.out > /dev/null 2>&1
