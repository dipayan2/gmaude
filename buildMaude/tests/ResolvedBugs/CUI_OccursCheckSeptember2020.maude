set show timing off .
set show advisories off .

***
***	Test for nontermination in U and CU unification due to
***	not doing an indirect occurs check. We use irred because
***	we might reduce the number of redundant unifiers in the
***	future.
***

fmod U-TEST is
  sort Foo .
  ops 1 a : -> Foo .
  op f : Foo Foo -> Foo [id: 1] .
vars A B C X Y Z : Foo .
endfm

irred unify X =? f(Z, Y) /\ Z =? f(X, Y) /\ f(X, Y) =? f(A, B) .
irred unify f(X, Y) =? a /\ Z =? f(X, Y) /\ X =? f(Z, Y)  .
irred unify X =? f(f(f(X, Y), X), Y) .
irred unify X =? f(f(f(X, X), Z), Y) .
irred unify X =? f(f(X, Y), f(Y, X)) .
irred unify X =? f(f(f(f(Y, X), Y), X), Z) .

fmod CU-TEST is
  sort Foo .
  ops 1 a : -> Foo .
  op f : Foo Foo -> Foo [comm id: 1] .
vars A B C X Y Z : Foo .
endfm

irred unify X =? f(Z, Y) /\ Z =? f(X, Y) /\ f(X, Y) =? f(A, B) .
irred unify f(X, Y) =? a /\ Z =? f(X, Y) /\ X =? f(Z, Y)  .
irred unify X =? f(f(f(X, Y), X), Y) .
irred unify X =? f(f(f(X, X), Z), Y) .
irred unify X =? f(f(f(f(Y, X), Y), X), Z) .
