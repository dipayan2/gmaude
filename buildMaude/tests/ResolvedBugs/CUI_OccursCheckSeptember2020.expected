==========================================
irredundant unify in U-TEST : X =? f(Z, Y) /\ Z =? f(X, Y) /\ f(X, Y) =? f(A,
    B) .

Unifier 1
X --> f(#1:Foo, #2:Foo)
Z --> f(#1:Foo, #2:Foo)
Y --> 1
A --> #1:Foo
B --> #2:Foo
==========================================
irredundant unify in U-TEST : f(X, Y) =? a /\ Z =? f(X, Y) /\ X =? f(Z, Y) .

Unifier 1
X --> a
Y --> 1
Z --> a
==========================================
irredundant unify in U-TEST : X =? f(f(f(X, Y), X), Y) .

Unifier 1
X --> 1
Y --> 1
==========================================
irredundant unify in U-TEST : X =? f(f(f(X, X), Z), Y) .

Unifier 1
X --> 1
Z --> 1
Y --> 1
==========================================
irredundant unify in U-TEST : X =? f(f(X, Y), f(Y, X)) .

Unifier 1
X --> 1
Y --> 1
==========================================
irredundant unify in U-TEST : X =? f(f(f(f(Y, X), Y), X), Z) .

Unifier 1
X --> 1
Y --> 1
Z --> 1
==========================================
irredundant unify in CU-TEST : X =? f(Z, Y) /\ Z =? f(X, Y) /\ f(X, Y) =? f(A,
    B) .

Unifier 1
X --> f(#1:Foo, #2:Foo)
Z --> f(#1:Foo, #2:Foo)
Y --> 1
A --> #1:Foo
B --> #2:Foo
==========================================
irredundant unify in CU-TEST : f(X, Y) =? a /\ Z =? f(X, Y) /\ X =? f(Z, Y) .

Unifier 1
X --> a
Y --> 1
Z --> a
==========================================
irredundant unify in CU-TEST : X =? f(f(f(X, Y), X), Y) .

Unifier 1
X --> 1
Y --> 1
==========================================
irredundant unify in CU-TEST : X =? f(f(f(X, X), Z), Y) .

Unifier 1
X --> 1
Y --> 1
Z --> 1
==========================================
irredundant unify in CU-TEST : X =? f(f(f(f(Y, X), Y), X), Z) .

Unifier 1
X --> 1
Z --> 1
Y --> 1
Bye.
